### PR TITLE
feat: role based permissions

### PR DESCRIPTION
**Pull request by liyiy**
## Changes
Initial implementation for roles

roles: `api/organizations/@current/roles`
role_memberships: `api/organizations/@current/roles/{role_id}/role_memberships`

New organizations created will have a default `Write` role associated with that org
New users joining a default org will have `Write` role memberships

"Write" access = our current default access

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

File: ee/api/role.py
Comment (neilkakkar) [inline ee/api/role.py:25]: oof I don't like `extract_organization` here!

Since we know `request.user` is a `User`, can we just do `request.user.organization` ?

`extract_organization` seems to have been created for cases where we don't know the model that its being used on.

File: ee/api/role.py
Comment (neilkakkar) [inline ee/api/role.py:75]: Why is `role_id` read_only? Don't you need this at creation time to know which role's membership you're creating?

File: ee/api/test/test_role_membership.py
Comment (neilkakkar) [inline ee/api/test/test_role_membership.py:31]: Also test here that the `RoleMembership` model has what you'd expect :)

File: posthog/models/organization.py
Comment (neilkakkar) [inline posthog/models/organization.py:72]: Q: Are organisations only created by using `bootstrap`, or is there some place where they're called by directly calling `create`?

File: ee/api/role.py
Comment (neilkakkar) [reply ee/api/role.py:75]: Ah, it's coming from the context, based on the url, I see!

File: ee/api/role.py
Comment (neilkakkar) [inline ee/api/role.py:77]: We should probably allow bulk updating here, as I imagine for big teams, it will be verrryyy annoying if they have to add members one by one to a role.

File: ee/api/role.py
Comment (neilkakkar) [inline ee/api/role.py:14]: Based on the mocks, I think admins can change the default role, right? (they'll need to make it 'read only' ).

I'm unclear on how this will change the model creation code below, since we can't imply permissions based on the role name 😅

**Review by neilkakkar — COMMENTED** [review]
Nice work!

One blocking comment (on default role name).

Rest are all just nits :)

File: ee/api/test/test_role.py
Comment (liyiy) [reply ee/api/test/test_role.py:86]: role object optional doesn't have name or something like that

File: posthog/models/user.py
Comment (neilkakkar) [inline posthog/models/user.py:224]: this sounds like a significant error?

File: posthog/models/organization.py
Comment (liyiy) [reply posthog/models/organization.py:72]: It looks like it's mostly "officially" created through `bootstrap` except for tests and internal metrics stuff? Maybe @Twixes knows more here too

File: ee/api/role.py
Comment (liyiy) [reply ee/api/role.py:14]: I think a way to do "read only" is by removing the write role, so an empty array of role permissions = view only. We don't have to add a `read` role but we can if it solves a problem that simply removing "write" doesn't?

**Comment by neilkakkar** [comment]
Idea: I think what might be useful to see before this PR goes in is the rest of the backend for how we'll apply these roles to feature flags (which should allow rapidly changing the models here if needed), as I'm sort-of unclear still how this will work without explicit permission levels 😬 .

File: posthog/models/organization.py
Comment (neilkakkar) [reply posthog/models/organization.py:72]: Not sure where `User.bootstrap` is called, but this might be problematic:

https://github.com/PostHog/posthog/blob/master/posthog/models/user.py#L70

Other that this, yeah, looks like it's just tests, or internal metrics.

Comment (liyiy) [reply posthog/models/organization.py:72]: ok I see it looks like that's when users sign up for the first time completely

File: ee/api/role.py
Comment (neilkakkar) [reply ee/api/role.py:14]: I see, so when an admin chooses to set default to 'read only', we remove every user in the org from `RoleMembership` to the `Write` role?

Comment (liyiy) [reply ee/api/role.py:14]: yepp if by default everyone is "read only" then we delete the write role memberships for everyone. And so team member A will have empty role memberships until new roles are created and they're added to it. And if the admin wants to change it back to default "write access" then we add that "write" role membership back to everyone.

File: posthog/models/organization.py
Comment (Twixes) [reply posthog/models/organization.py:72]: Hmm, I don't quite see why we need a default role. It seems cumbersome to ensure all orgs have a default role and that every person belongs to that role, when we could just implicitly defer to the organization/project's access settings. In my mind, a role would usually mean a real-world team (e.g. "Experimentation") or a role within a company ("Project Manager"), so a vague "Write" role just doesn't sit well with me… Is there some more context for this architectural choice?

File: ee/api/role.py
Comment (liyiy) [reply ee/api/role.py:14]: So then we only have to check if any role membership exists or not for a user to determine "view only" mode for them. If a user doesn't have any role memberships then we don't even need to check for feature flags access for them !

File: posthog/models/user.py
Comment (liyiy) [reply posthog/models/user.py:216]: > Hmm, I don't quite see why we need a default role. It seems cumbersome to ensure all orgs have a default role and that every person belongs to that role, when we could just implicitly defer to the organization/project's access settings. In my mind, a role would usually mean a real-world team (e.g. "Experimentation") or a role within a company ("Project Manager"), so a vague "Write" role just doesn't sit well with me… Is there some more context for this architectural choice?

How would you differentiate a full edit access vs view only access for a regular team member then? We can rename the role if that's the concern, I just used the example name from launchdarkly's "default roles". But how do you think we would determine "default full edit access" for everyone?

File: posthog/models/organization.py
Comment (neilkakkar) [reply posthog/models/organization.py:72]: @liyiy 's comment:

How would you differentiate a full edit access vs view only access for a regular team member then? We can rename the role if that's the concern, I just used the example name from launchdarkly's "default roles". But how do you think we would determine "default full edit access" for everyone?

---

I quite like this idea, tbh. Let default permission come from org settings (which determines what the default is - full access or view only), which are overridden by any roles if found for the given user.

File: ee/api/role.py
Comment (neilkakkar) [reply ee/api/role.py:14]: This is a bit problematic from a DB design perspective: https://en.wikipedia.org/wiki/Database_normalization

In particular, (yes I had to read that link again to come up with the right terms for things), there seems to be a transitive dependency with the role & permissions.

Disregard the default case for now since that might change, but consider a role like: `Schrodinger`, and every team member is a part of this role.

Every day, a cat opens up PostHog and walks on the keyboard, such that the permissions for this role toggle from `full write access` to `view only`. One day it's `view only`, the next day it's `full write access`, and then back, and so on.

Based on the above design, when we switch to `view only` we update `RoleMembership` to remove all users from the role. And same for when we switch to `full write access`, we add them back.

1. This in unnecessary updates to the DB, since changing permissions shouldn't affect `RoleMembership`, which can lead to issues with DB management, and hard to get right (since all rows to be updated need to be in an atomic transaction, etc. etc)
2. How do we figure out who was in the role, when permissions change? How do we add back only the people who had this role?

Comment (liyiy) [reply ee/api/role.py:14]: 1. Wouldn't updating all users to have a "read only" role by patching their role membership also require an atomic transaction? 
2. This makes sense I think? Can you explain how does having a default role that you can change the name of answer these problems? 

What do you suggest to do here instead?

File: posthog/models/organization.py
Comment (liyiy) [reply posthog/models/organization.py:72]: okay, so by default being a member+ will grant you edit access. And then if you want to restrict certain members to view only, then you add a "view only" membership?

File: ee/api/role.py
Comment (EDsCODE) [inline ee/api/role.py:40]: We'll want to return `feature_flag_access_level` here too

File: ee/api/role.py
Comment (liyiy) [reply ee/api/role.py:74]: Bulk create/updates on the API is not that intuitive here as you can see with the `.bulk_create` I've been testing with, I see us mostly doing it from the frontend, check the API calls + logic for `dashboard_collaborators` for something similar

Comment (liyiy) [reply ee/api/role.py:74]: Will delete the commented out code below for bulk creation in another commit unless there's valid reasons why we want to try to make bulk creation work on the API end

File: posthog/models/organization.py
Comment (liyiy) [inline posthog/models/organization.py:37]: This is causing some gnarly circular imports for me when trying to use `Organization` in the ee `Role` model, happy to take suggestions here if there's a better way about this than importing the `License` only when we need it for this file

**Comment by EDsCODE** [comment]
One of the remaining questions for backend is if we’re using OganizationResourceAccess for Orgs, should we do a similar pattern for role?

**Comment by liyiy** [comment]
> One of the remaining questions for backend is if we’re using OganizationResourceAccess for Orgs, should we do a similar pattern for role?

I don't think we need to right now unless it solves for something? This sounds like more work to parse through than its benefits provide currently. I think `OrganizationResourceAccess` provides a very clear benefit of detaching it from non ee `organization` along with preventing future updates to that table's fields for the different `access levels` because organization is super intertwined with many things. And if we do find ourselves needing to migrate to a `RoleResourceAccess` pattern I imagine the deprecation for `feature_flags_access_level` on it + the transaction to create the existing `role resource access` model instances wouldn't be too bad? I'm open to suggestions though.

File: ee/api/organization_resource_access.py
Comment (neilkakkar) [inline ee/api/organization_resource_access.py:17]: Looks like you'll need a filter on the organisation as well here?

Comment (neilkakkar) [reply ee/api/organization_resource_access.py:17]: (also test with multiple orgs having the same resource types)

File: ee/api/organization_resource_access.py
Comment (neilkakkar) [inline ee/api/organization_resource_access.py:41]: Can potentially be n+1 queries if this fetches multiple orgs. Should test to confirm 😅  - I'm not sure if the org filter is applied here by the `StructuredViewSetMixin`.

File: ee/api/feature_flag_role_access.py
Comment (neilkakkar) [inline ee/api/feature_flag_role_access.py:17]: same, filter on org here!

(unless I'm missing something and this happens automatically)

Comment (liyiy) [reply ee/api/feature_flag_role_access.py:17]: I think this happens automatically but I'll double check!

File: ee/api/feature_flag_role_access.py
Comment (neilkakkar) [inline ee/api/feature_flag_role_access.py:32]: `all()` seems unnecessary here?

```suggestion
            request.user.role_memberships
```

File: ee/api/feature_flag_role_access.py
Comment (neilkakkar) [inline ee/api/feature_flag_role_access.py:36]: minor nit

```suggestion
        return has_role_membership_with_access
```

File: ee/api/role.py
Comment (neilkakkar) [reply ee/api/role.py:74]: Works

File: ee/urls.py
Comment (neilkakkar) [inline ee/urls.py:37]: where is this different router coming from? 🤔

File: ee/models/organization_resource_access.py
Comment (neilkakkar) [inline ee/models/organization_resource_access.py:12]: what does this mean?

File: ee/api/test/test_feature_flag_role_access.py
Comment (neilkakkar) [inline ee/api/test/test_feature_flag_role_access.py:24]: shouldn't this be `ONLY_VIEW` for this test to make sense?

Comment (neilkakkar) [reply ee/api/test/test_feature_flag_role_access.py:24]: i.e. this person can ONLY_VIEW flags, except for the flags they created, where they can add other roles.

File: ee/api/test/test_feature_flag_role_access.py
Comment (neilkakkar) [inline ee/api/test/test_feature_flag_role_access.py:52]:
```suggestion
        self.assertEqual(response_data, self.permission_denied_response("You can't edit roles for this feature flag."))
```

File: ee/api/test/test_organization.py
Comment (neilkakkar) [inline ee/api/test/test_organization.py:207]: this is going to be annoying for whoever needs to work on this later 😅 - as we are doing the revert only when the test is successful, vs. being in a context manager / beforeEach-afterEach block.

Will look into the `Roles` circular imports tomorrow 👀

File: ee/api/test/test_organization_resource_access.py
Comment (neilkakkar) [inline ee/api/test/test_organization_resource_access.py:9]: can remove this

File: ee/api/test/test_organization_resource_access.py
Comment (neilkakkar) [inline ee/api/test/test_organization_resource_access.py:36]: would be nice to have multiple orgs in this test

File: ee/api/test/test_role.py
Comment (neilkakkar) [inline ee/api/test/test_role.py:61]: same, multiple orgs in test 😅

**Review by neilkakkar — COMMENTED** [review]
Nice job with the tests!

File: ee/api/test/test_organization.py
Comment (liyiy) [reply ee/api/test/test_organization.py:207]: yeaaaa i didn't know how else to get around this, would def like some suggestions here !

**Comment by neilkakkar** [comment]
Floating again as this might've gotten lost somewhere:

Can we get rid of `DEFAULT_VIEW_ALLOW_EDIT_BASED_ON_ROLE` ? I'm not sure what this achieves 🤔

File: ee/models/organization_resource_access.py
Comment (liyiy) [reply ee/models/organization_resource_access.py:12]: Members can be view only or edit only and then we won't need to check if they belong to any feature flag designated roles. Otherwise that level denotes "custom feature flag role" access which is for when you want the user to be view only for everything but have access to certain flags. It communicates more clearly to customers that they can set strictly "view only" or "edit always" or a third custom "per flag access" for their users and roles. 

Additionally we now only have to check if a user belongs to the "custom access" role before checking what feature flags x roles they're connected to instead of always having to do that with "view only" permissions, so it should save us an extra search

Comment (liyiy) [reply ee/models/organization_resource_access.py]: removed

File: ee/api/test/test_role.py
Comment (liyiy) [reply ee/api/test/test_role.py:59]: tested

File: ee/api/test/test_organization_resource_access.py
Comment (liyiy) [reply ee/api/test/test_organization_resource_access.py:38]: ✅

File: ee/api/organization_resource_access.py
Comment (liyiy) [reply ee/api/organization_resource_access.py:41]: ✅

File: ee/migrations/0014_roles_memberships_and_resource_access.py
Comment (neilkakkar) [inline ee/migrations/0014_roles_memberships_and_resource_access.py:13]: is this right? The other file shows latest as 280_

**Review by neilkakkar — APPROVED** [review]
One comment, and one failing test but otherwise LGTM! :)

File: ee/migrations/0014_roles_memberships_and_resource_access.py
Comment (liyiy) [reply ee/migrations/0014_roles_memberships_and_resource_access.py]: good catch !

**Pull request merged**
